### ISSUE TITLE
WebSocketClient not removing closed WebSocket Session's from managed beans

### ISSUE THREAD

**Issue by bistvan** [eclipse/jetty.project#2655]
Hi,

when a Websocket Session is closed, but instead of closing the WebSocketClient, a new websocket is opened, the annotated Websocket Handler is kept in the memory.

Please find the attached java example (put a breakpoint at line "System.out.println("Put a breakpoint here")". When it stops the client.toString() will show 100 open sessions, but they are already closed.

[JettyClientCloseBug.zip](https://github.com/eclipse/jetty.project/files/2094297/JettyClientCloseBug.zip)

**Comment by joakime** [comment]
Can replicate.

Client Sessions are not being removed from WebSocketClient beans properly.

**Comment by joakime** [comment]
Merged into jetty-9.4.x

**Comment by menulis** [comment]
Websocket sessions (and their corresponding WebSocketClient threads) terminated due to idle timeout do not seem to be properly destroyed as of 9.4.14.v20181114. Implicitly calling close() on javax.websocket.Session inside Endpoint's onError/onClose or elsewhere does not help (though I assume this should not be necessary as Session.isOpen already reports false).

**Comment by menulis** [comment]
@joakime Is it possible to reopen this bug as per my report one month ago?

**Comment by sbordet** [comment]
@menulis can you try the latest code on branch `jetty-9.4.x`, or the staged release at https://oss.sonatype.org/content/repositories/jetty-1451?

**Comment by menulis** [comment]
Thanks for the reply, @sbordet. I tried https://oss.sonatype.org/content/repositories/jetty-1451 and I no longer see the problem of `JsrSession` object buildup inside the `ClientContainer`, eventually leading to OOM, which is great.

With respect to the increasing number of threads with every newly established session, this is still the case if sessions are esablished by `ContainerProvider.getWebSocketContainer().connectToServer()`, that is, for every new session, a new `WebSocketContainer` instance is created (unless `JettyClientContainerProvider.willUseSingleton(true)` is called?). The workaround is to either call `stop()` on (casted) `WebSocketClient` when closing the session or call `ContainerProvider.getWebSocketContainer()` only once and use it for all subsequent connectToServer() calls.

Do you have any estimates about when the next release will be available?

**Comment by joakime** [comment]
eclipse-ee4j/websocket-api#212 - From a purely JSR356 point of view, there is no lifecycle for a JSR356 Client Container.  It is supposed to live for the entirety of the JVM.

`JettyClientContainerProvider.willUseSingleton(true)` is a hack within Jetty to allow each call to `ContainerProvider.getWebSocketContainer()` to return a single instance, vs how the JSR356 API says it will return a new container with each call.

In regards to `stop()`, that's a Jetty [`LifeCycle`](https://www.eclipse.org/jetty/javadoc/current/org/eclipse/jetty/util/component/LifeCycle.html) behavior.
If you have any other Jetty components that are an instance of [`ContainerLifeCycle`](https://www.eclipse.org/jetty/javadoc/current/org/eclipse/jetty/util/component/ContainerLifeCycle.html) then you can simply do [`thatContainerComponent.addBean(websocketclientinstance);`](https://www.eclipse.org/jetty/javadoc/current/org/eclipse/jetty/util/component/ContainerLifeCycle.html#addBean-java.lang.Object-) and then that container component will automatically stop the websocket client when that container component stops.

Example:
https://github.com/eclipse/jetty.project/blob/8973f01dc4935339743715898ab9dc9e894da5b5/jetty-websocket/javax-websocket-client-impl/src/test/java/org/eclipse/jetty/websocket/jsr356/CookiesTest.java#L108-L109

**Comment by sbordet** [comment]
@menulis your usage of `WebSocketContainer` is wrong: you should not create an instance for every attempt to connect to the server.
You create one instance, and then reuse that multiple times to connect to the server.

Regarding stopping the `WebSocketContainer`, you can follow what @joakime suggests, or this simpler approach:

```java
WebSocketContainer container = ContainerProvider.getWebSocketContainer();
container.connectToServer(...);

// When done with the container:
LifeCycle.stop(container);
```

**Comment by menulis** [comment]
Thanks for clarifications, @sbordet and @joakime, they make a lot of sense and I have already adjusted the code accordingly. Stopping the container upon the shutdown of JVM is in fact a minor problem when compared to the memory leak due to `JsrSession` object buildup inside the `ClientContainer` and I am glad this is fixed in the staging release [9.4.15.v2019021](https://oss.sonatype.org/content/repositories/jetty-1451). Do you have any estimates of when it will be pushed to the next public release?

**Comment by sbordet** [comment]
@menulis Jetty 9.4.15 has landed in Maven Central, so it's now available to the public. Thanks for your feedback on this issue!
